Add unit tests for pkg/datapath/loader

joestringer opened the pull request:
Based upon ~~https://github.com/cilium/cilium/pull/6180~~ .

```
$ sudo -E make tests-privileged TESTPKGS=github.com/cilium/cilium/pkg/datapath/loader
can't load package: package github.com/cilium/cilium/test/bpf: C source files not allowed when not using cgo or SWIG: bpf-event-test.c unit-test.c
Starting key-value store containers...                                              
cilium-etcd-test-container                                                      
c82049bf4918b79499a1d11acbcc24a01099743c2d823b52255b18b17e857928                   
cilium-consul-test-container                                                
aa9893473cfc45ed563a3c94a274955a3c71dca63a53a6d5ba18f08cef7e0dd0          
  CHECK contrib/scripts/bindata.sh                                                 
?       github.com/cilium/cilium/pkg/datapath/loader    [no test files]
=== RUN   Test                               
START: loader_test.go:123: LoaderTestSuite.TestCompileAndLoad                   
PASS: loader_test.go:123: LoaderTestSuite.TestCompileAndLoad    1.609s         
                                                                                
START: loader_test.go:149: LoaderTestSuite.TestCompileFailure               
level=error msg="Command execution failed" cmd="[llc -march=bpf -mcpu=probe -filetype=obj -o /tmp/cilium_814020965/bpf_lxc.o]" error="signal: killed" subsys=datapath-loader
level=error msg="Failed to compile bpf_lxc.o: signal: killed" compiler-pid=14494 linker-pid=14496 subsys=datapath-loader
level=warning msg="JoinEP: Failed to compile" debug=false error="Failed to compile bpf_lxc.o: signal: killed" params="&{Source:bpf_lxc.c Output:bpf_lxc.o OutputType:obj}" subsys=datapath-loader
PASS: loader_test.go:149: LoaderTestSuite.TestCompileFailure    0.103s
                                                                                  
START: loader_test.go:132: LoaderTestSuite.TestReload            
PASS: loader_test.go:132: LoaderTestSuite.TestReload    3.637s
                                                                                
OK: 3 passed                                    
--- PASS: Test (5.35s)                             
PASS                                                                   
coverage: 75.8% of statements                                                      
ok      github.com/cilium/cilium/pkg/datapath/loader    5.403s  coverage: 75.8% of statements
cilium-etcd-test-container                                                
cilium-consul-test-container
```

Most of the remaining coverage is error cases which are hard to properly test.

The test `TestCompileFailure` is expected to fail, hence why there are error logs but the test is marked as passed in the above example output.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6117)
<!-- Reviewable:end -->

joestringer committed:
loader: Mark test as privileged

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
loader: Move device preparation to TestMain

Create a new TestMain function for setup of the Cilium device once so
that it isn't recreated for each unit test run.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
loader: Differentiate state input vs output

This had the potential for upcoming tests to accidentally overwrite the
source files under bpf/, because the input source directory was the same
as the output directory. Split these.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
loader: Add privileged unit tests for compile,load

Signed-off-by: Joe Stringer <joe@cilium.io>

aanm commented on pkg/datapath/loader/loader_test.go:
```diff
@@ -11,14 +11,15 @@
 // limitations under the License.
-
-// +build !privileged_tests
+//
```
Remove the `//`

joestringer replied:
```suggestion

```

aanm commented on pkg/datapath/loader/loader_test.go:
```diff
@@ -11,14 +11,15 @@
-
-// +build !privileged_tests
+//
+// +build privileged_tests
```
Out of curiosity, why does it need to be privileged?

joestringer replied:
Because it's creating devices and loading programs onto them. At the very least the `bpf()` syscall requires root privileges for what we're doing here.

aanm replied:
@joestringer I should have been more clear. The remove `//` was to be replaced with a new line. Just that the copyright is detached from the `build`

joestringer replied:
I thought that's what I did, but I did it via GH's new change suggestion tool which did this instead. Sigh.

Anyway, what difference does it make?

aanm replied:
`+build` might be considered part of the copyright header when modifying the copyright header with automatic tools

joestringer replied:
I fixed this up in the new version.

joestringer commented on pkg/datapath/loader/loader_test.go:
```diff
@@ -35,12 +35,43 @@ type LoaderTestSuite struct{}
+	var result int
+
+	// Defer exit to the end to ensure that subsequent cleanup occurs
+	defer os.Exit(result)
```
I intended this to pick up the value of result at the time of deferral, but it appears to pick it up at this point. Should shuffle this function around to avoid the defers and have everything execute in the appropriate order.

joestringer committed:
loader: Fail replaceDatapath on migrate failure

If the `cilium-map-migrate` binary cannot be found or fails in some way,
this should prevent the replaceDatapath() from returning successfully.

Fixes: 168572f84f04 ("loader: replace join endpoint script with loader")
Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
loader: Inject include dependencies at test time

The default includes sometimes do not include this path, which causes
upcoming tests to fail. Add an extra include path to account for this,
which is only populated properly from the unit tests.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
test: Make map migration available in unit tests

Build the bpf/ directory and export a PATH including that directory.
This allows the `cilium-map-migrate` binary under the bpf/ directory to
be found by unit tests run via this docker-compose file.

Signed-off-by: Joe Stringer <joe@cilium.io>

ianvernon commented on ginkgo.Jenkinsfile:
```diff
@@ -53,6 +53,24 @@ pipeline {
                }
             }
         }
+        stage('UnitTesting') {
```
Rename this stage to `Privileged Unit Testing` and add a comment in the `Jenkinsfile` indicating why we are running this stage in Jenkins vs. in Travis.

ianvernon replied:
This applies more to 
https://github.com/cilium/cilium/pull/6180 - I'll leave this comment on there.

joestringer commented on pkg/datapath/loader/loader_test.go:
```diff
@@ -35,12 +36,48 @@ type LoaderTestSuite struct{}
+	}
+	defer os.RemoveAll(tmpDir)
+	if dirInfo, err = getDirs(tmpDir); err != nil {
+		log.Fatal(err.Error())
```
Convert to return (int, error) and put the `log.Fatal()` in the caller to ensure that the deferred cleanup runs.

joestringer commented on pkg/datapath/loader/loader_test.go:
```diff
@@ -35,12 +36,48 @@ type LoaderTestSuite struct{}
+func runTests(m *testing.M) int {
+	testIncludes = "-I/usr/include/x86_64-linux-gnu/"
+
+	tmpDir, err := ioutil.TempDir("/tmp/", "cilium_")
```
Replace `/tmp/` with `""` to use the default directory, and make the prefix more specific to this test.
